lxd/storage_lvm.go

```diff
@@ -759,31 +759,69 @@ func (s *storageLvm) ContainerMount(name string, path string) (bool, error) {
 	mountOptions := s.volume.Config["block.mount_options"]
 	containerMntPoint := getContainerMountPoint(s.pool.Name, name)
 
-	if shared.IsMountPoint(containerMntPoint) {
+	containerMountLockID := fmt.Sprintf("mount/%s/%s", s.pool.Name, name)
+	lxdStorageLock.Lock()
+	if waitChannel, ok := lxdStorageLockMap[containerMountLockID]; ok {
+		lxdStorageLock.Unlock()
+		if _, ok := <-waitChannel; ok {
+			shared.LogWarnf("Value transmitted over image lock semaphore?")
+		}
+		// Give the benefit of the doubt and assume that the other
+		// thread actually succeeded in mounting the storage volume.
 		return false, nil
 	}
 
-	err := tryMount(containerLvmPath, containerMntPoint, lvFsType, 0, mountOptions)
-	if err != nil {
-		return false, fmt.Errorf("Error mounting snapshot LV path='%s': %s", path, err)
+	lxdStorageLockMap[containerMountLockID] = make(chan bool)
+	var imgerr error
+	ourMount := false
+	if !shared.IsMountPoint(containerMntPoint) {
+		imgerr = tryMount(containerLvmPath, containerMntPoint, lvFsType, 0, mountOptions)
+		ourMount = true
+	}
+	if waitChannel, ok := lxdStorageLockMap[containerMountLockID]; ok {
+		close(waitChannel)
+		delete(lxdStorageLockMap, containerMountLockID)
+	}
+	lxdStorageLock.Unlock()
```

Comment:
You're holding that lock for a long time here (can be up to 10s if tryMount fails)